server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -658,14 +664,54 @@ public void write(byte[] b, int offset, int length) throws IOException {
         }
 
         private void sendNextChunk(long position, BytesArray content, boolean lastChunk) throws IOException {
-            // Actually send the file chunk to the target node, waiting for it to complete
-            cancellableThreads.executeIO(() ->
+            // Actually send the file chunk to the target node.
+            cancellableThreads.executeIO(() -> {
+                sendingTickets.acquire();
+                try {
+                    throwAndClearErrorIfExist();
                     recoveryTarget.writeFileChunk(md, position, content, lastChunk, translogOps.get())
-            );
+                        .whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                synchronized (this) {
+                                    if (error == null) {
+                                        error = (Exception) ex;
+                                    } else {
+                                        error.addSuppressed(ex);
+                                    }
+                                }
+                            }
+                            sendingTickets.release();
+                        });
+                } catch (Exception e) {
+                    sendingTickets.release();
+                    throw e;
+                }
+            });
             if (shard.state() == IndexShardState.CLOSED) { // check if the shard got closed on us
                 throw new IndexShardClosedException(request.shardId());
             }
         }
+
+        @Override
+        public void close() {
+            // wait for the completion of all ongoing chunks then check for the existing error.
+            cancellableThreads.execute(() -> sendingTickets.acquire(maxConcurrentFileChunks));
+            try {
+                throwAndClearErrorIfExist();
+            } finally {
+                sendingTickets.release(maxConcurrentFileChunks);
+            }
+        }
+
+        private void throwAndClearErrorIfExist() {
```

Comment:
do we need the "clearing" part? I think there's no need to set this back to null